### ISSUE TITLE
Error response from daemon: Unexpected http code: 400 when working with Harbor

### ISSUE THREAD

**Issue by loopif** [vmware/vic#3441]
As a user of VIC, I should be able to specify ":443" during installation with vic-machine.

As a user of VIC, I should be able to pull from a public library (project in Harbor) in a secure registry if I installed the VCH with --insecure-registry.

Acceptance Criteria:  Robot scripts to validate vic-machine's URL validation for insecure-registry.  Robot scripts to validate the ability to pull an image from a public library from a private Harbor registry using TLS.

---

VCH (engine 0.8.0-rc3) shows "Error response from daemon: Unexpected http code: 400, URL: http://192.168.2.44:443/v2/" when PULLing image from local Harbor registry (secured by TLS, accessed through https://192.168.2.44) by docker -H 192.168.2.32:2376 --tls pull 192.168.2.44:443/library/photon:1.0. 

VCH is created with vic-machine with --insecure-registry 192.168.2.44:443. 

There is NO error when PULLing from client-machine (not from VCH but Red Hat) dockerd daemon (v. 1.12.3, client API set to 1.23 for VIC host ) started with the same --insecure-registry 192.168.2.44:443. What's a problem?

**Comment by mdubya66** [comment]
@sflxn could you take a look?

**Comment by loopif** [comment]
Harbor v. 0.5.0-rc2 is used. Should I post steps to reproduce the error?

**Comment by sflxn** [comment]
@loopif There is a validation issue up there.  Something we need to add.  You specify insecure-registry (http), but you specified port 443 (https).  It's a mismatch.  I'm not sure that's the cause of the error you're getting, but it's the first thing I noticed.

Did you docker login <harbor> before pulling?

**Comment by loopif** [comment]
I've noticed that too. Here is what I found.
1. On my Red Hat, _ps -fp PID_ shows _/usr/bin/dockerd --insecure-registry 192.168.2.44:443_ because I start dockerd with _--insecure-registry 192.168.2.44:443_. 
Subsequently, _docker info_ for my Red Hat shows 

> Registry: https://index.docker.io/v1/
> Insecure Registries:
>  192.168.2.44:443
>  127.0.0.0/8

 I can login and pull without log in if it's public repository like default library/photon:1.0

2. I create VCH: _vic-machine-linux create --target 'administrator@vsphere.local':'passwd'@vcenter.lab.local/Datacenter --name vic_host --thumbprint 2D:7F:DC:14:20:C8:47:59:C9:C5:3C:0D:F6:98:C4:37:6C:E3:F3:53 --image-store Main_Datastore --volume-store Main_Datastore/vic:default --bridge-network VIC_DPG --name vic_host **--insecure-registry 192.168.2.44:443** --no-tlsverify_

However! I SSH to the VCH and run _ps -fp PID_ which shows _/sbin/docker-engine-server -port=2376 -port-layer-port=2377_ without _--insecure-registry 192.168.2.44:443_ specified above when creating a VCH 

_docker -H 192.168.2.121:2376 --tls info_ only shows this 

> Registry: registry-1.docker.io

without Insecure Registries which means that vic-machine doesn't provide the docker daemon with _--insecure-registry_

**Comment by sflxn** [comment]
@loopif I've been able to verify your procedure.

1) You found a bug.  It looks like we're not handling :443 correctly.  You shouldn't ever have to specify :443.  We do need to check that.
2) You're installing the VCH incorrectly and you're using docker's insecure-registry incorrectly

Insecure-registry should only be used if you want to use the registry WITHOUT TLS.  You mentioned in the original post that you're installing Harbor WITH TLS.

Here's instructions on how to setup Harbor, VCH and use it with TLS.

https://github.com/vmware/vic/tree/master/doc/application_workflow/harbor_registry

**Comment by sflxn** [comment]
One last comment, you say that you had no problems pulling from Harbor using standard docker daemon.  Unless you copied the CA cert from Harbor, you should have had problems.  Try docker logout of the harbor server and attempt to log back in using standard docker.  You should get an error.  The link I provided above walks you through this.

**Comment by sflxn** [comment]
There is one more thing you need to know about container images and registries.  If the library is set to public, you do not need to login.  I suspect this is why you didn't have any problems when pulling with standard docker.  If the library is set to private, you must login.  In this scenario, standard docker should have given you an error.

If you log into Harbor, there are options to create "projects".  In Docker's terminology, that's a library.  Once you create a "project", you can set the publicity of that library.  If you read through my example in the doc link I sent you, I tagged the images as

<harbor ip>/test/busybox

"test" is the project.  If you tag the image as <harbor ip>/busybox, that specify the default library, which is public, and you will not need to login.

Hate to dump so much info onto you, but registries, pulling, and login has all these subtleties.  I'll soon do a video demonstrating all these subtleties.

**Comment by loopif** [comment]
I believe that I set up correctly according to documentation. A few more words that might help troubleshoot.

You said 

> Basically, you need --registry-ca parameter for TLS. --insecure-registry is used for registries not using TLS.
...
Unless you copied the CA cert from Harbor, you should have had problems.

_--insecure-registry_ should NOT only mean NO TLS at all. 

According to [official Docker doc.](https://docs.docker.com/engine/reference/commandline/dockerd/#/insecure-registries), 

> A secure registry uses TLS and a copy of its CA certificate is placed on the Docker host at /etc/docker/certs.d/myregistry:5000/ca.crt. An insecure registry is either not using TLS (i.e., listening on plain text HTTP), or is **using TLS with a CA certificate not known by the Docker daemon**. **The latter can happen when the certificate was not found under /etc/docker/certs.d/myregistry:5000/, or if the certificate verification failed (i.e., wrong CA).** 

In my case, Harbor is using TLS but VCH is not aware of Harbor's CA certificate. So VCH should still work _with --insecure-registry_.

This is why my local Docker daemon started with _--insecure-registry_  is unaware of Harbor's CA **(I haven't imported and put a CA cert. in a right directory according to documentation)** and still able to pull, login without the error.

To sum up: my local Docker CAN login, push, pull from public and private repositories on Harbor with NO problem. A conclusion is that VCH's docker daemon gets configured incorrectly.

**Comment by loopif** [comment]
But I can confirm that it works OK with the same setup with only replaced _--insecure-registry_ to _--registry-ca=ca.crt_ from Harbor

**Comment by sflxn** [comment]
@loopif Yes, I agree with what you posted from Docker.  We will actually attempt https before http for insecure-registry.  Your attempt should have given you a different error message if it cannot pull down the image.  We have a missing validation for ':443'.  That's what's causing the http 400 status.  If you had specified --insecure-registry with just the IP and no ':443', it would have given you the other warning.

As for local docker with Harbor and insecure-registry, this is new information to me.  I thought that Harbor sets up to run as HTTP or HTTPS not both.  That gets set in the nginx config file.  I'll go verify this with the Harbor team.

**Comment by sflxn** [comment]
I just got confirmation from the Harbor team that it should only respond to HTTP or HTTPS but not both.  Just adding insecure-registry to standard docker without updating it with the CA cert should not have worked.  You should get an error:

```
x509: certificate signed by unknown authority
```

**Comment by sflxn** [comment]
I re-examined this issue again today in order to answer @stuclem's question about documentation.  There were a few issues in rolled up in this issue.

1) Lack of validating :443 in vic-machine.  This does not appear to be an issue anymore.  The latest version, 0.9.0 results in
```
Mar  8 2017 14:40:20.351-08:00 ERROR vic-machine-linux create failed: 10.118.69.162:443 is an invalid format for registry url
```
2) VIC cannot pull from a secure registry if we specify --insecure-registry during installation.  This IS an issue.  I just spent some time verifying it is a bug.  When @loopif reported this issue, I was stuck on issue 1, but yes, it's an issue.

Moving this to backlog to fix in the next sprint.

**Comment by sflxn** [comment]
@stuclem If you care about documentation before the next sprint's code gets merged, you can say

Currently, VIC always performs certificate verification with a secure registry even when --insecure-registry is specified during installation with vic-machine.  This is a known issue.

**Comment by stuclem** [comment]
Thanks @sflxn. I added the following in the release notes:

------

- <b>`Error response from daemon: Unexpected http code: 400` when pulling images from local Harbor registries. [#3441](https://github.com/vmware/vic/issues/3441)</b><br> Currently, vSphere Integrated Containers Engine always performs certificate verification with a secure registry even if you specify `vic-machine create --insecure-registry` during deployment of the VCH. VMware is working to resolve the issue.


-----

**Comment by mdubya66** [comment]
making this high priority for inclusion in the next sprint.
